components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt

```diff
@@ -0,0 +1,168 @@
+package mozilla.components.feature.downloads.manager
+
+import android.Manifest.permission.INTERNET
+import android.Manifest.permission.WRITE_EXTERNAL_STORAGE
+import android.annotation.TargetApi
+import android.content.Context
+import android.net.Uri
+import android.os.Build
+import android.os.Build.VERSION.SDK_INT
+import android.os.Environment
+import androidx.core.content.getSystemService
+import androidx.core.net.toUri
+import kotlinx.coroutines.Dispatchers.IO
+import kotlinx.coroutines.withContext
+import mozilla.components.browser.session.Download
+import mozilla.components.concept.fetch.Client
+import mozilla.components.concept.fetch.Headers.Names.CONTENT_LENGTH
+import mozilla.components.concept.fetch.Headers.Names.CONTENT_TYPE
+import mozilla.components.concept.fetch.MutableHeaders
+import mozilla.components.concept.fetch.Request
+import mozilla.components.concept.fetch.Response
+import mozilla.components.support.ktx.android.content.isPermissionGranted
+import java.io.File
+import java.io.FileOutputStream
+import java.io.OutputStream
+
+class FetchDownloadManager(
+    private val applicationContext: Context,
+    private val httpClient: Client
+) : DownloadManager {
+
+    override suspend fun download(download: Download, refererUrl: String, cookie: String): CompletedDownload? {
+        if (download.isSupportedProtocol()) {
+            // We are ignoring everything that is not http or https. This is a limitation of
+            // GeckoView: https://bugzilla.mozilla.org/show_bug.cgi?id=1501735 and
+            // https://bugzilla.mozilla.org/show_bug.cgi?id=1432949
+            return null
+        }
+
+        if (!applicationContext.isPermissionGranted(INTERNET, WRITE_EXTERNAL_STORAGE)) {
+            throw SecurityException("You must be granted INTERNET and WRITE_EXTERNAL_STORAGE permissions")
+        }
+
+        val headers = MutableHeaders().apply {
+            safeAppend(CONTENT_TYPE, download.contentType)
+            safeAppend("User-Agent", download.userAgent)
+            safeAppend("Cookie", cookie)
```

Comment:
`Cookie` and `User-Agent` primarily exist for the WebView flow and for Android's Download Manager:
* We can omit `User-Agent` since Gecko(View) already uses its own (We only need to tell Android's Download Manager to use one that matches the browser)
* We can set a `cookiePolicy` on the request to let Gecko(View) use whatever cookies it has for this request. This again only exists to pass of cookies from WebView to Android's Download Manager. This should never be set by Gecko(View) and in the worst case we are overriding the Cookie header here with an empty string.

Reply:
I think we should keep the `Cookie` and `User-Agent` headers as the consumers of `FetchDownloadManager` could pass whatever children of `Client` not necessarily `GeckoViewFetchClient`. In the specific case of `GeckoViewFetchClient`, we could check if the `cookie` are empty in that case Gecko(View) could use whatever cookies it has if the `cookiePolicy` specified that.